### DISCUSSION TITLE
Differentiate the validation failure based on a severity level

### DISCUSSION THREAD

**Discussion by pmad7954** [networknt/json-schema-validator#745] [Q&A]
Hi Team, 

First I need to tell you that I'm very satisfied with the json-schema-validator library and I appreciate your contribution there.

If I explain my question, simply I need to differentiate the validation failures, errors, and warns, and based on the type I decide how to print that through a Java code. I raise my question in StackOverflow but I did not get any answers there. I hope someone here will help me to do what I need. I have explained this thoroughly in the  StackOverflow question, use the following link to view it.

[Stackoverflow question](https://stackoverflow.com/questions/76202046/add-aditional-information-to-json-schema-and-get-it-when-validate-through-the-ne)

TIA!

**Comment by fdutton** [comment]
@pmad7954 We already have an open issue (#680) for a similar feature and as noted by @stevehu, we currently expect the provided schema to be valid and only do minimal testing of the schema. Right now, the library only produces errors about the JSON instance being validated; it does not produce any other type of message.

I also have a need to validate my schemas but I cannot do that until all the _schema keywords_ work as specified and pass all the JSON Schema Test Suite conformance tests.

I assume that your warnings would be equivalent to the lint output mentioned in that issue. I have a few other ideas on this:
- Validating a schema would produce an error when encountering a malformed _keyword_. For example, `prefixItems` requires an array of schema so I should get an error when the value is not an array.
- Validating a schema would produce a warning when encountering a _keyword_ that has no effect at that location. For example, attempting to use `prefixItems` on an `object` would produce a warning since that _keyword_ only applies to `array`. The specification says to ignore the _keyword_ in this case but I think it would be useful for schema authors to have this information.
- Trying to reference a remote schema that results in an IO error currently results in an exception being raised. I would find it useful to report it as a fatal message rather than raising an exception but this behavior would be controlled by a configuration flag.

Can you provide examples of what you would consider an error and warning in this context?

**Comment by fdutton** [comment]
@pmad7954 I just read you Stackoverflow question and that is a little different from what you described here. If you only need programmatic access to the extra information, you could try leveraging the Custom Message support to hack something. See
https://github.com/networknt/json-schema-validator/blob/master/doc/cust-msg.md

**Reply by pmad7954** [reply]
There can be minor differences @fdutton, but my expectation is the same, which is to use an additional property to hold the information I need. Anyway, I use the custom message feature from the beginning for the validation schemas which is very helpful.

**Comment by fdutton** [comment]
I posted a response to your Stackoverflow question.

I think your question was overlooked because there are many libraries that qualify as `json-schema-validator` and they exist in many languages such as Java, JavaScript and .Net. I think we need a tag that is more specific to this library.

**Reply by pmad7954** [reply]
Thank you @fdutton, I checked that. I already implemented that workaround until a long-term solution implements.

**Comment by visheshruparelia** [comment]
Hi @fdutton / @pmad7954 , I also have a similar requirement (I want to be able to categorise the validation failures and only if the failure is of a certain severity I want to act on it, otherwise just ignore it). While I do see how we can use custom messages to solve this problem as mentioned in the stack overflow answer, but it can be limiting for my exact requirement. I am wondering if [JsonSchemaWalkers](https://github.com/networknt/json-schema-validator/blob/master/doc/walkers.md#json-schema-walkers) can be somehow used here? What do you think?

**Reply by fdutton** [reply]
@visheshruparelia The walker functionality is primarily targeted at traversing the schema and not the instance (i.e., the data) being validated. If you need supplemental validation of the schema then this approach could work. It would be a bigger effort to insert this approach into the normal validation functionality.

One of the issues with assigning severity when validating a JSON instance is that each client will have a different opinion on what validations constitute an error, a warning or simply informational. In this case, it would be better if the client filters the validation results based on the error code. While this library assigns an error code to each keyword, it does not expose this information in the validation messages, unfortunately. Additionally, the format messages lack this attribute.

**Reply by visheshruparelia** [reply]
Thanks @fdutton for the response. 
> The walker functionality is primarily targeted at traversing the schema and not the instance (i.e., the data) being validated.

Got it, not applicable to my usecase then. 
> While this library assigns an error code to each keyword, it does not expose this information in the validation messages, unfortunately.

Are there any plans to make this available in validation messages in future? Also can you point me to the code which does this error assignment?

The reason I am bit reluctant in taking the custom messages approach is because the Schema against which I am gonna run validation will be publicly available. So if I add `message` attribute in the schema itself, folks who are going to create instances of that schema will be able to see that - this can lead to unnecessary confusion because that extra field is only added for internal consumption. Is there any another approach which I can take? Or perhaps add the `message` fields programmatically at runtime while validating (this has its own set of challenges but at least it won't be exposed to external consumers of the schema).

**Reply by fdutton** [reply]
@visheshruparelia The error codes are the second parameter to the [ValidatorTypeCode](https://github.com/networknt/json-schema-validator/blob/master/src/main/java/com/networknt/schema/ValidatorTypeCode.java) enum.
